microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md

```diff
@@ -0,0 +1,314 @@
+
+### Node types
+
+The first number found in the group of numbers for a node in the `nodes` array corresponds to its type. 
+
+This number is an index that can be used to lookup the type name in the `snapshot.meta.node_types[0]` array.
+
+| Node type | Description |
+| --- | --- |
+| Hidden | A V8 internal element that doesn't correspond directly to a user-controllable JavaScript object. In DevTools, these all show up under the category name **(system)**. Even though these objects are internal, they can be an important part of retainer paths. |
+| Object | Any user-defined object like `{ x: 2 }` or `new Foo(4)`. Contexts, which show up in DevTools as **system / Context** hold variables that had to be allocated on the heap because they're used by a nested function. |
+| Native | Things allocated by the Blink rendering engine, not by V8. These are mostly DOM: HTMLDivElement, CSSStyleRule, etc. |
+| Concatenated string | The result of concatenating two strings with the `+` operator. Rather than making a new string with a copy of all the data from the two source strings, V8 creates a ConsString object with pointers to the two source strings. From a JavaScript perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
+| Sliced string | The result of a substring operation, such as using `String.prototype.substr` or `String.prototype.substring`. V8 avoids copying string data by instead creating a SlicedString which points to the original string and specifies the start index and length. From a JS perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
+| Array | Various internal lists, which show up in the devtools with the category name **(array)**. Like Hidden, this category clumps together a bunch of unrelated things. Many of the objects here are named **(object properties)** or **(object elements)**, indicating that they contain the string-keyed or numeric-keyed properties of a JavaScript object. |
+| Code | Things that grow proportionally to the amount of script, and/or the number of times that functions run. |
+| Synthetic | Synthetic nodes don't correspond to anything actually allocated in memory. These are used to distinguish the different kinds of GC roots. |
+
+### Edges
+
+Similar to the `nodes` array, the `edges` top-level array contains all the individual elements needed to recreate the edges of the graph.
+
+Also similar to nodes, the total number of edges can be calculated by multiplying `snapshot.edge_count` with `snapshot.meta.edge_fields.length`. Edges are also stored as a sequence of numbers which you will need to iterate on by groups of `snapshot.meta.edge_fields.length`.
+
+However, to read the `edges` array correctly, you first need read the `nodes` array because each node knows how many edges it has. 
```

Comment:
```suggestion
However, to read the `edges` array correctly, you first need to read the `nodes` array, because each node knows how many edges it has. 
```